### ISSUE TITLE
The training of BaseLayer pauses at all_to_all_single function

### ISSUE THREAD

**Issue by ZhiYuanZeng** [pytorch/fairseq#3730]
### To Reproduce

Steps to reproduce the behavior (**always include the command you ran**):

1. Run cmd
```shell
python -m torch.distributed.launch \
    --nnodes=1 \
    --nproc_per_node=8 \
    --master_addr="127.0.0.1" \
    --master_port=12345 \
    $(which fairseq-train) ${data_dir} \
    --distributed-world-size 8 \
    --arch transformer \
    --base-layers 1 \
    --optimizer adam \
    --max-tokens 500 \
    --max-epoch 30
```
2. See the training process pause at the 5th step
![image](https://user-images.githubusercontent.com/50764547/126647551-3a44e840-440f-4fdd-bfc8-b280459c0eb2.png)

### Environment

 - fairseq Version: https://github.com/pytorch/fairseq/tree/eff39d5d453497a5a6e5e998e2a920fb5f0618e1
 - PyTorch Version: 1.9
 - OS (e.g., Linux): Ubuntu
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): `pip install --editable ./`
 - Python version: 3.6.1
 - CUDA/cuDNN version: 10.1.105

### Aditional Information
I have found that the processes block at the `all_to_all_single` function. But I can not find the problem.

**Comment by Hanlard** [comment]
I have the same problem

**Comment by ZhiYuanZeng** [comment]
I have found some cases that may cause the problem:
1. The shape of the tensor in different devices is not the same. Fairseq collects a batch according to the sentence length dynamically, therefore different devices have different batch sizes. The solution is to pad the tensor to be the same shape.
2. The number of tokens in a  batch can not be divided by the number of devices. The solution is also padding, and make sure the padded shape is fine.
3. During translating, some processes may finish, while others are still running. The running processes are waiting at the all2all communication since the processes group contains all processes, including the finished ones. The solution is to let the finished processes do some dummy translation until all processes are finished

**Comment by Hanlard** [comment]
It seems that the 'torch.distributed.all_to_all_single' operator has no limit to the input tensor shape or the tokens number...

**Comment by Hanlard** [comment]
> I have found some cases that may cause the problem:
> 
> 1. The shape of the tensor in different devices is not the same. Fairseq collects a batch according to the sentence length dynamically, therefore different devices have different batch sizes. The solution is to pad the tensor to be the same shape.
> 2. The number of tokens in a  batch can not be divided by the number of devices. The solution is also padding, and make sure the padded shape is fine.
> 3. During translating, some processes may finish, while others are still running. The running processes are waiting at the all2all communication since the processes group contains all processes, including the finished ones. The solution is to let the finished processes do some dummy translation until all processes are finished

It seems that the 'torch.distributed.all_to_all_single' operator has no limit to the input tensor shape or the tokens number...

**Comment by ZhiYuanZeng** [comment]
It seems that if the input tensor of different devices is not the same, the input_splits and output_splits should be given.

> It seems that the 'torch.distributed.all_to_all_single' operator has no limit to the input tensor shape or the tokens number...
